airbyte-integrations/connectors/source-avni/source_avni/source.py

```diff
@@ -0,0 +1,180 @@
+
+        yield from data
+
+    def update_state(self) -> None:
+
+        if self.last_record:
+            updated_last_date = self.last_record["audit"]["Last modified at"]
+            self.state = {self.cursor_field[1]: updated_last_date}
+            self.last_record = None
+
+        return None
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+
+        total_elements = int(response.json()["totalElements"])
+        page_size = int(response.json()["pageSize"])
+
+        if total_elements == page_size:
+            self.current_page = self.current_page + 1
+            return {"page": self.current_page}
+
+        self.update_state()
+
+        self.current_page = 0
+
+        return None
+
+
+class IncrementalAvniStream(AvniStream, IncrementalMixin, ABC):
+
+    state_checkpoint_interval = None
+
+    @property
+    def cursor_field(self) -> List[str]:
+        return ["audit", "Last modified at"]
+
+    @property
+    def state(self) -> Mapping[str, Any]:
+
+        if self.cursor_value:
+            return {self.cursor_field[1]: self.cursor_value}
+        else:
+            return {self.cursor_field[1]: self.lastModifiedDateTime}
+
+    @state.setter
+    def state(self, value: Mapping[str, Any]):
+        self.cursor_value = value[self.cursor_field[1]]
+        self._state = value
+
+
+class Subjects(IncrementalAvniStream):
+    def path(self, **kwargs) -> str:
+        return "subjects"
+
+
+class ProgramEnrolments(IncrementalAvniStream):
+    def path(self, **kwargs) -> str:
+        return "programEnrolments"
+
+
+class ProgramEncounters(IncrementalAvniStream):
+    def path(self, **kwargs) -> str:
+        return "programEncounters"
+
+
+class Encounters(IncrementalAvniStream):
+    def path(self, **kwargs) -> str:
+        return "encounters"
+
+
+class SourceAvni(AbstractSource):
+    def get_client_id(self):
+
+        url_client = "https://app.avniproject.org/idp-details"
+        response = requests.get(url_client)
+        response.raise_for_status()
+        client = response.json()
+        return client["cognito"]["clientId"]
+
+    def get_token(self, username: str, password: str, app_client_id: str) -> str:
+
+        client = boto3.client("cognito-idp", region_name="ap-south-1")
+        response = client.initiate_auth(
+            ClientId=app_client_id, AuthFlow="USER_PASSWORD_AUTH", AuthParameters={"USERNAME": username, "PASSWORD": password}
+        )
+        return response["AuthenticationResult"]["IdToken"]
+
+    def check_connection(self, logger, config) -> Tuple[bool, any]:
+
+        username = config["username"]
+        password = config["password"]
+
+        try:
+            client_id = self.get_client_id()
+        except Exception as error:
+            return False, str(error) + ": Please connect With Avni Team"
+
+        try:
+            auth_token = self.get_token(username, password, client_id)
+        except Exception as error:
```

Comment:
Check https://discuss.airbyte.io/t/how-the-airbyte-cdk-check-command-works/4589 how to implement the check function properly